README.md
## pelican-extended-sitemap

A sitemap plugin for [Pelican](https://github.com/getpelican/pelican).

It generates a sitemap.xml according to the specification of [sitemaps.org](http://sitemaps.org) and considers the pelican index page, articles, pages and special pages (like tags, categories, authors).
Also comes with a XML stylesheet to be able to view the sitemap in browser without pain.

**ATTENTION: Plugin is in heavy development - do not use yet!**

### ToDos

What still has to be implemented:

* make a real python package
* i18n (maybe)


### Pelican settings

Add to the plugins list:


```python
PLUGINS = [
    'plugins.extended-sitemap'
]
```

### Plugin settings

Add the `EXTENDED_SITEMAP_PLUGIN` dict to your settings.
The keys explained:

* priorities: priority for each page type, from 0.0 to 1.0
 * index: index page
 * articles: article pages
 * pages: pages
 * others: category, tags and authors pages
* changefrequencies: how often a page will likely change, possible values: always, hourly, daily, weekly, monthly, yearly, never

The settings below are the default values:

```python
EXTENDED_SITEMAP_PLUGIN = {
    'priorities': {
        'index': 1.0,
        'articles': 0.8,
        'pages': 0.5,
        'others': 0.4
    },
    'changefrequencies': {
        'index': 'daily',
        'articles': 'weekly',
        'pages': 'monthly',
        'others': 'monthly',
    }
}
```


CHANGELOG
1.0.2
-----

* if the following settings are not set, no sitemap entry is generated for the appropriate pages
    * CATEGORY_URL
    * TAG_URL
    * AUTHOR_URL

1.0.1
-----

* fixed compatibility with Pelican 3.5
* dates are now W3C compliant and do not include time
* adjusted setup.py

1.0.0
-----

* fixed wrong package naming, now "extended_sitemap" instead "extended-sitemap"
* added unit tests
* raise ConfigurationError if TIMEZONE setting is not set
* tags are now sorted by name

0.2.0
-----

* Python 3 support

0.1.4
-----

* fixed date format in sitemap
* changed xml stylesheet link for this package to PyPI

0.1.3
-----

* adjusted documentation

0.1.2
-----

* adjusted documentation

0.1.1
-----

* adjusted documentation

0.1.0
-----

* initial release covering index, articles, pages, categories, tags and author pages



extended_sitemap/tests/__init__.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, print_function

import filecmp
import locale
import os
import subprocess
import unittest

from extended_sitemap import ConfigurationError

from tempfile import mkdtemp

from pelican import Pelican
from pelican.settings import read_settings
from pelican.tests.support import mute


# used paths
CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
CONTENT_DIR = os.path.abspath(os.path.join(CURRENT_DIR, 'content'))
EXPECTED_DIR = os.path.abspath(os.path.join(CURRENT_DIR, 'expected'))
OUTPUT_PATH = os.path.abspath(os.path.join(CURRENT_DIR, 'output'))


class FileComparisonTest(unittest.TestCase):
    """
    Unittest class with possibility to assert equal file contents.
    """
    def assertFileContentEquals(self, path_file_expected, path_file_test):
        """
        Asserts the file contents to be equal.
        :param path_file_expected: path to the file with the expected content
        :type path_file_expected: str
        :param path_file_test: path to the file to test
        :type path_file_test: str
        """
        if not filecmp.cmp(path_file_expected, path_file_test):
            msg_fail = 'File content of %(filename)s does not match expected content!' % {'filename': path_file_test}

            # if there is git and git diff works for both files, append the file diff to the fail message
            try:
                out, err = subprocess.Popen(
                    ['git', 'diff', '--minimal', '--no-color', '--no-ext-diff', '--exit-code', '-w', path_file_expected, path_file_test],
                    stdout=subprocess.PIPE,
                    stderr=subprocess.PIPE,
                ).communicate()
                if len(err) == 0:
                    msg_fail += '\n\n' + out.decode('utf-8')
            except OSError:
                # if there is no git, just don't output the diff
                pass

            self.fail(msg_fail)


class ExtendedSitemapTest(FileComparisonTest):

    def setUp(self):
        self.path_temp = mkdtemp(prefix='extended_sitemap_tests.')
        self.path_cache = mkdtemp(prefix='extended_sitemap_cache.')

        # default minimal configuration for pelican in test context
        self.settings_default = {
            'PAGE_DIR': os.path.join(CONTENT_DIR, 'pages'),
            'ARTICLE_DIR': os.path.join(CONTENT_DIR, 'articles'),
            'PATH': CONTENT_DIR,
            'OUTPUT_PATH': self.path_temp,
            'CACHE_PATH': self.path_cache,
            'LOCALE': locale.normalize('en_US'),
            'SITEURL': 'http://example.com',
            'PLUGIN_PATH': os.path.join(CURRENT_DIR, '..'),
            'PLUGINS': ['extended_sitemap'],
        }

    def __execute_pelican(self, settings_override=None):
        """
        Executes pelican. Uses the minimal config of self.settings_default (that will fail!) merged with the given additional settings.
        :param settings_override: dictionary with pelican setting values to set
        :type settings_override: dict
        """
        if not settings_override:
            settings_override = {}

        settings = self.settings_default.copy()
        settings.update(settings_override)

        pelican_settings = read_settings(
            path=None,
            override=settings
        )
        pelican = Pelican(settings=pelican_settings)
        mute(True)(pelican.run)()

    def test_timezone_missing(self):
        """
        As the TIMEZONE settings is necessary to create timezone based date value, ensure the configuration exception is raised if it is not configured.
        """
        self.assertRaises(ConfigurationError, self.__execute_pelican)

    def test_sitemap_structure(self):
        """
        Tests basic structure of generated sitemap.
        """
        self.__execute_pelican(
            settings_override={
                'TIMEZONE': 'Europe/Berlin',
            }
        )
        self.assertFileContentEquals(
            os.path.join(EXPECTED_DIR, 'test_sitemap_structure.xml'),
            os.path.join(self.path_temp, 'sitemap.xml')
        )

    def test_sitemap_structure_subpaths(self):
        """
        Tests basic structure of generated sitemap with subpath in domain.
        """
        # issue #2
        self.__execute_pelican(
            settings_override={
                'TIMEZONE': 'Europe/Berlin',
                'SITEURL': 'http://example.com/subpath',
            }
        )
        self.assertFileContentEquals(
            os.path.join(EXPECTED_DIR, 'test_sitemap_structure_subpath.xml'),
            os.path.join(self.path_temp, 'sitemap.xml')
        )


extended_sitemap/__init__.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, print_function

import datetime
import os
import sys

from codecs import open

from pelican import signals

from pytz import timezone

if sys.version_info >= (3, 0):
    from urllib.parse import urljoin
else:
    from urlparse import urljoin


class ConfigurationError(Exception):
    """
    Exception class for wrong configurations.
    """
    pass


class SitemapGenerator(object):
    """
    Class for generating a sitemap.xml.
    """

    xml_wrap = """<?xml version="1.0" encoding="UTF-8"?><?xml-stylesheet type="text/xsl" href="%(SITEURL)ssitemap-stylesheet.xsl"?>
<urlset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd" xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
%(urls)s
</urlset>"""

    template_url = """<url>
<loc>%(loc)s</loc>
<lastmod>%(lastmod)s</lastmod>
<changefreq>%(changefreq)s</changefreq>
<priority>%(priority).2f</priority>
</url>"""

    settings_default = {
        'priorities': {
            'index': 1.0,
            'articles': 0.8,
            'pages': 0.5,
            'others': 0.4
        },
        'changefrequencies': {
            'index': 'daily',
            'articles': 'weekly',
            'pages': 'monthly',
            'others': 'monthly',
        }
    }

    def __init__(self, context, settings, path, theme, output_path, **kwargs):
        """
        Initializes the generator class.
        :param context: the generated context, mix of settings and transformed content
        :type context: dict
        :param settings: the pelican project settings
        :type settings: dict
        :param path: the path to the content files
        :type path: str
        :param theme: the path to the theme
        :type theme: str
        :param output_path: the path where the generated output is put
        :type output_path: str
        :param kwargs: additional keyword arguments
        :type kwargs: dict
        """
        self.path_content = path
        self.path_output = output_path
        self.context = context
        if settings.get('TIMEZONE', None) is None:
            raise ConfigurationError('Please specify the TIMEZONE setting!')
        self.timezone = timezone(settings.get('TIMEZONE'))
        self.url_site = settings.get('SITEURL')
        # Pelican strips off trailing slashes during settings initialization.
        # The later used urljoin function strips of path elements not ending with a trailing slash,
        # a slash is added here if it is not already present
        if not self.url_site.endswith('/'):
            self.url_site += '/'
        self.settings = settings.get('EXTENDED_SITEMAP_PLUGIN', self.settings_default)

    def generate_output(self, writer):
        """
        Generates the sitemap file and the stylesheet file and puts them into the content dir.
        :param writer: the writer instance
        :type writer: pelican.writers.Writer
        """
        # write xml stylesheet
        with open(os.path.join(os.path.dirname(__file__), 'sitemap-stylesheet.xsl'), 'r', encoding='utf-8') as fd_origin:
            with open(os.path.join(self.path_output, 'sitemap-stylesheet.xsl'), 'w', encoding='utf-8') as fd_destination:
                xsl = fd_origin.read()
                # replace some template markers
                # TODO use pelican template magic
                xsl = xsl.replace('{{ SITENAME }}', self.context.get('SITENAME'))
                fd_destination.write(xsl)

        # will contain the url nodes as text
        urls = ''

        # get all articles sorted by time
        articles_sorted = sorted(self.context['articles'], key=lambda x: x.date, reverse=True)

        # the landing page
        if 'index' in self.context.get('DIRECT_TEMPLATES'):
            # assume that the index page has changed with the most current article
            urls += self.__create_url_node_for_content(
                articles_sorted[0],
                'index',
                url=self.url_site,
            )

        # process articles
        for article in articles_sorted:
            urls += self.__create_url_node_for_content(article, 'articles')

        # process pages
        for page in sorted(self.context.get('pages'), key=lambda x: x.date, reverse=True):
            urls += self.__create_url_node_for_content(
                page,
                'pages',
                url=urljoin(self.url_site, page.url)
            )

        # process category pages
        if self.context.get('CATEGORY_URL'):
            urls += self.__process_url_wrapper_elements(self.context.get('categories'))

        # process tag pages
        if self.context.get('TAG_URL'):
            urls += self.__process_url_wrapper_elements(sorted(self.context.get('tags'), key=lambda x: x[0].name))

        # process author pages
        if self.context.get('AUTHOR_URL'):
            urls += self.__process_url_wrapper_elements(self.context.get('authors'))

        # write the final sitemap file
        with open(os.path.join(self.path_output, 'sitemap.xml'), 'w', encoding='utf-8') as fd:
            fd.write(self.xml_wrap % {
                'SITEURL': self.url_site,
                'urls': urls
            })

    def __process_url_wrapper_elements(self, elements):
        """
        Creates the url nodes for pelican.urlwrappers.Category and pelican.urlwrappers.Tag.
        :param elements: list of wrapper elements
        :type elements: list
        :return: the processes urls as HTML
        :rtype: str
        """
        urls = ''
        for url_wrapper, articles in elements:
            urls += self.__create_url_node_for_content(
                url_wrapper,
                'others',
                url=urljoin(self.url_site, url_wrapper.url),
                modification_time=sorted(articles, key=lambda x: x.date, reverse=True)[0].date
            )
        return urls

    def __create_url_node_for_content(self, content, content_type, url=None, modification_time=None):
        """
        Creates the required <url> node for the sitemap xml.
        :param content: the content class to handle
        :type content: pelican.contents.Content
        :param content_type: the type of the given content to match settings.EXTENDED_SITEMAP_PLUGIN
        :type content_type; str
        :param url; if given, the URL to use instead of the url of the content instance
        :type url: str
        :param modification_time: the modification time of the url, will be used instead of content date if given
        :type modification_time: datetime.datetime
        :returns: the text node
        :rtype: str
        """
        return self.template_url % {
            'loc': url if url is not None else urljoin(self.url_site, self.context.get('ARTICLE_URL').format(**content.url_format)),
            'lastmod': modification_time.strftime('%Y-%m-%d') if modification_time is not None else content.date.strftime('%Y-%m-%d'),
            'changefreq': self.settings.get('changefrequencies').get(content_type),
            'priority': self.settings.get('priorities').get(content_type),
        }


def get_generators(generators):
    """
    Returns the generators of this plugin,
    :param generators: current generators
    :type generators: pelican.Pelican
    :returns: the sitemap generator type
    :rtype: type
    """
    return SitemapGenerator


def register():
    """
    Registers the sitemap generator.
    """
    signals.get_generators.connect(get_generators)
